[Task #3] Write tests for all data services

Aleksander-Drumev-WA opened the pull request:
Catalog and Basket data services integration tests created.

Aleksander-Drumev-WA committed:
Adding test project.
Catalog and Basket data services integration tests.

Aleksander-Drumev-WA committed:
Combining fixture class files.

var1ap-WA commented on Pizzeria.Tests/UserBrowsingCatalogItemsTests.cs:
```diff
@@ -0,0 +1,76 @@
+namespace Pizzeria.Tests
+{
+    [Collection("Database collection")]
+    public class UserBrowsingCatalogItemsTests
```
rename pls to `CatalogDataServiceTests`

Aleksander-Drumev-WA replied:
Done.

var1ap-WA commented on Pizzeria.Tests/UserBrowsingCatalogItemsTests.cs:
```diff
@@ -0,0 +1,76 @@
+        public async Task Create_And_Get_All_Catalog_Items()
+        {
+            // Arrange
+            var fixture = new DatabaseFixture();
```
move to helper

Aleksander-Drumev-WA replied:
Moved.

var1ap-WA commented on Pizzeria.Tests/UserBrowsingCatalogItemsTests.cs:
```diff
@@ -0,0 +1,76 @@
+                .Should()
+                .NotBeEmpty()
+                .And
+                .HaveCount(4);
```
You need to add more check for at least for one of catalog item (name,price,...)

Aleksander-Drumev-WA replied:
Done.

var1ap-WA commented on WA.Pizza.Infrastructure/Data/Services/CatalogDataService.cs:
```diff
@@ -19,20 +19,23 @@ public CatalogDataService(AppDbContext dbContext)
+
+            var catalogItem = dto.Adapt<CatalogItem>();
+
+            await _dbContext.CatalogItems.AddAsync(catalogItem);
```
Read about `AddAsync` and `Add`

Aleksander-Drumev-WA replied:
Done

var1ap-WA commented on WA.Pizza.Infrastructure/Data/Services/CatalogDataService.cs:
```diff
@@ -19,20 +19,23 @@ public CatalogDataService(AppDbContext dbContext)
 
         // functionality for pagination?
-        public ICollection<ListCatalogItemsDTO> GetAllAsync()
+        public ICollection<ListCatalogItemsDTO> GetAll()
```
rewrite to Async version

Aleksander-Drumev-WA replied:
Done.

var1ap-WA commented on Pizzeria.Tests/Fixtures/DatabaseFixture.cs:
```diff
@@ -0,0 +1,38 @@
+            _connection = new SqlConnection("Server=DESKTOP-5CAFU3F\\SQLEXPRESS;Database=WA.PizzaDB.Tests;Trusted_Connection=True;");
+            _options = new DbContextOptionsBuilder<AppDbContext>().UseSqlServer(_connection.ConnectionString).Options;
+            DbContext = new AppDbContext(_options);
+            DbContext.Database.Migrate();
```
add call to mapping config

Aleksander-Drumev-WA replied:
Done.

var1ap-WA commented on Pizzeria.Tests/UserBasketInteractionTests.cs:
```diff
@@ -0,0 +1,346 @@
+namespace Pizzeria.Tests
+{
+    [Collection("Database collection")]
+    public class UserBasketInteractionTests
```
rename pls

Aleksander-Drumev-WA replied:
Renamed.

var1ap-WA commented on Pizzeria.Tests/UserBasketInteractionTests.cs:
```diff
@@ -0,0 +1,346 @@
+        public async Task Anonymous_user_adds_catalog_item_to_basket()
+        {
+            // Arrange
+            var fixture = new DatabaseFixture();
```
same

Aleksander-Drumev-WA replied:
It's moved.

var1ap-WA commented on Pizzeria.Tests/UserBasketInteractionTests.cs:
```diff
@@ -0,0 +1,346 @@
+            var result = await sut.AddItemToBasketAsync(null, request);
+
+            // Assert
+            result.Should().Be(1);
```
bad check

Aleksander-Drumev-WA replied:
It's better now.

var1ap-WA commented on Pizzeria.Tests/UserBasketInteractionTests.cs:
```diff
@@ -0,0 +1,346 @@
+            var result = await sut.AddItemToBasketAsync(user.Id, request);
+
+            // Assert
+            result.Should().Be(1);
```
bad assert

Aleksander-Drumev-WA replied:
It's better now.

var1ap-WA commented on Pizzeria.Tests/UserBasketInteractionTests.cs:
```diff
@@ -0,0 +1,346 @@
+            await fixture.DbContext.Baskets.AddAsync(basket);
+            await fixture.DbContext.SaveChangesAsync();
+
+            foreach (var catalogItem in catalogItems)
+            {
+                var request = new CatalogItemToBasketItemRequest
+                {
+                    BasketId = basket.Id,
+                    CatalogItemId = catalogItem.Id,
+                    Quantity = 5
+                };
+
+                await sut.AddItemToBasketAsync(user.Id, request);
+            }
```
rewrite it pls in easier way

Aleksander-Drumev-WA replied:
Rewritten.

var1ap-WA commented on Pizzeria.Tests/UserBasketInteractionTests.cs:
```diff
@@ -0,0 +1,346 @@
+            var result = sut.GetBasketWithBasketItems(basket.Id);
+
+            // Assert
+            result.Select(b => b.BasketItems.Should().HaveCount(4));
```
not good assert

Aleksander-Drumev-WA replied:
It's better now.

var1ap-WA commented on Pizzeria.Tests/UserBasketInteractionTests.cs:
```diff
@@ -0,0 +1,346 @@
+                PasswordHash = Lorem.Sentence(8),
+            };
+            await fixture.DbContext.Users.AddAsync(user);
+            var catalogItems = new List<CatalogItem>
+            {
+                new CatalogItem
+                 {
+                     Name = Lorem.Sentence(5),
+                     Price = RandomNumber.Next(1, int.MaxValue),
+                     PictureBytes = Lorem.Sentence(5),
+                     StorageQuantity = 25
+                 },
+                 new CatalogItem
+                 {
+                     Name = Lorem.Sentence(5),
+                     Price = RandomNumber.Next(1, int.MaxValue),
+                     PictureBytes = Lorem.Sentence(5),
+                     StorageQuantity = 25
+                 },
+                 new CatalogItem
+                 {
+                     Name = Lorem.Sentence(5),
+                     Price = RandomNumber.Next(1, int.MaxValue),
+                     PictureBytes = Lorem.Sentence(5),
+                     StorageQuantity = 25
+                 },
+                 new CatalogItem
+                 {
+                     Name = Lorem.Sentence(5),
+                     Price = RandomNumber.Next(1, int.MaxValue),
+                     PictureBytes = Lorem.Sentence(5),
+                     StorageQuantity = 25
+                 }
+            };
```
move to helper

Aleksander-Drumev-WA replied:
Moved.

var1ap-WA commented on Pizzeria.Tests/UserBasketInteractionTests.cs:
```diff
@@ -0,0 +1,346 @@
+            await fixture.DbContext.Baskets.AddAsync(basket);
+            await fixture.DbContext.SaveChangesAsync();
+
+            foreach (var catalogItem in catalogItems)
+            {
+                var request = new CatalogItemToBasketItemRequest
+                {
+                    BasketId = basket.Id,
+                    CatalogItemId = catalogItem.Id,
+                    Quantity = 5
+                };
+
+                await sut.AddItemToBasketAsync(user.Id, request);
+            }
```
rewrite

Aleksander-Drumev-WA replied:
Rewritten.

var1ap-WA commented on Pizzeria.Tests/UserBasketInteractionTests.cs:
```diff
@@ -0,0 +1,346 @@
+            };
+
+            // Act
+            var result = await sut.UpdateItemAsync(dto);
```
rewrite method to return ID

Aleksander-Drumev-WA replied:
Done

var1ap-WA commented on Pizzeria.Tests/UserBasketInteractionTests.cs:
```diff
@@ -0,0 +1,346 @@
+            }
+
+            // Act
+            await sut.RemoveBasketItemAsync(3);
```
3 is magic number

Aleksander-Drumev-WA replied:
It's  changed

var1ap-WA commented on WA.Pizza.Infrastructure/Data/Services/BasketDataService.cs:
```diff
@@ -61,7 +70,7 @@ public async Task<int> UpdateItemAsync(BasketItemDTO updatedBasketItem)
             await _dbContext.SaveChangesAsync();
 
 
-            return localBasketItem.Id;
+            return localBasketItem.Quantity;
```
Big mistake

Aleksander-Drumev-WA replied:
Edited

Aleksander-Drumev-WA committed:
Helper method for creating objects created.
One instance of db is used across all tests.
Combining Add and Update methods for CatalogItem.
A bit better assertions.

var1ap-WA commented on WA.Pizza.Infrastructure/Services/Mapster/MappingConfig.cs:
```diff
@@ -29,18 +29,19 @@ public static void Configure()
-				.Map(dest => dest.Name, src => src.Name)
-				.Map(dest => dest.Price, src => src.Price.ToString("0.00"))
-				.Map(dest => dest.PictureBytes, src => src.PictureBytes);
+			//TypeAdapterConfig<CatalogItem, CatalogItemDTO>
+			//	.NewConfig()
+			//	.Ignore(dest => dest.StorageQuantity)
+			//	.Map(dest => dest.Id, src => src.Id)
+			//	.Map(dest => dest.Name, src => src.Name)
+			//	.Map(dest => dest.Price, src => src.Price.ToString("0.00"))
+			//	.Map(dest => dest.PictureBytes, src => src.PictureBytes);
```
Why?

Aleksander-Drumev-WA replied:
I combined Add and Update in CatalogDataService so I don't use this mapping anymore.

var1ap-WA commented on Pizzeria.Tests/BasketDataServiceTests.cs:
```diff
@@ -0,0 +1,244 @@
+            // Act
+            Func<Task> act = () => sut.AddItemToBasketAsync(null, request);
+
+            // Act, Assert
```
only Assert

Aleksander-Drumev-WA replied:
Now it is.

var1ap-WA commented on Pizzeria.Tests/BasketDataServiceTests.cs:
```diff
@@ -0,0 +1,244 @@
+                BasketItems = basketItems
+            };
+            await _fixture.DbContext.Baskets.AddAsync(basket);
+            await _fixture.DbContext.SaveChangesAsync();
```
var catalogItem = catalogItems.First();

Aleksander-Drumev-WA replied:
Done, but not in all tests. I need indexation in some tests because we don't want to add the same CatalogItem multiple times as BasketItem.

var1ap-WA commented on Pizzeria.Tests/BasketDataServiceTests.cs:
```diff
@@ -0,0 +1,244 @@
+            await sut.RemoveBasketItemAsync(basketItems[3].Id);
+
+            // Assert
+            basket.BasketItems.Should().HaveCount(3).And.NotContain(x => x.Id == 4);
```
check in DB

Aleksander-Drumev-WA replied:
Now I'm checking it

Aleksander-Drumev-WA committed:
BasketItem domain model changed.
Better assertions.
OrderDataService tests written.

var1ap-WA commented on Pizzeria.Tests/BasketDataServiceTests.cs:
```diff
@@ -19,39 +19,42 @@ namespace Pizzeria.Tests
                 .Should()
                 .HaveCount(1)
                 .And
                 .Contain(x => x.Quantity == request.Quantity);
```
What about  other fields?

Aleksander-Drumev-WA replied:
Now Included

var1ap-WA commented on Pizzeria.Tests/BasketDataServiceTests.cs:
```diff
@@ -120,36 +127,38 @@ public async Task User_checks_his_basket_with_basket_items()
+            result.Should().NotBeEmpty();
             result.Select(b => b.BasketItems.Should().HaveCount(4));
-            basketItemToAssert.Quantity.Should().Be(5);
+            result.Select(b => b.BasketItems.All(bi => bi.Quantity == 5).Should().BeTrue());
```
5 is magic number here

Aleksander-Drumev-WA replied:
Fixed

var1ap-WA commented on Pizzeria.Tests/BasketDataServiceTests.cs:
```diff
@@ -162,42 +171,44 @@ public async Task User_changes_basketItem_quantity()
             var basketItems = new List<BasketItem>();
-            for (int i = 0; i < 4; i++)
+
+            for (int i = 0; i < catalogItems.Count; i++)
```
why not `foreach` by catalogItems?

Aleksander-Drumev-WA replied:
All for loops are replaced with foreach now.

var1ap-WA commented on Pizzeria.Tests/BasketDataServiceTests.cs:
```diff
@@ -162,42 +171,44 @@ public async Task User_changes_basketItem_quantity()
-                PictureBytes = catalogItems[2].PictureBytes,
-                Price = catalogItems[2].Price,
+                Name = catalogItems[0].Name,
+                Price = catalogItems[0].Price,
```
var firstCatalogItem = catalogItems.First()

Aleksander-Drumev-WA replied:
Done

var1ap-WA commented on Pizzeria.Tests/CatalogDataServiceTests.cs:
```diff
@@ -17,38 +17,34 @@ namespace Pizzeria.Tests
 
-            // Assert
-            var catalogItemToAssert = await _fixture.DbContext.CatalogItems.FindAsync(catalogItems[1].Id);
+            // Assert graph comparison
```
delete pls `graph comparison`

Aleksander-Drumev-WA replied:
Done

var1ap-WA commented on Pizzeria.Tests/OrderDataServiceTests.cs:
```diff
@@ -0,0 +1,203 @@
+
+            // Assert
+            var orderToAssert = await _dbContext.Orders.FindAsync(orderId);
+            orderToAssert.OrderItems.Should().NotBeNull();
```
`orderToAssert.Should().NotBeNull();`

Aleksander-Drumev-WA replied:
Done.

var1ap-WA commented on Pizzeria.Tests/OrderDataServiceTests.cs:
```diff
@@ -0,0 +1,203 @@
+            // Assert
+            var orderToAssert = await _dbContext.Orders.FindAsync(orderId);
+            orderToAssert.OrderItems.Should().NotBeNull();
+            orderToAssert.OrderItems.Should().HaveCount(catalogItems.Count);
```
https://fluentassertions.com/objectgraphs/

Aleksander-Drumev-WA replied:
I think I did it the right way now

var1ap-WA commented on Pizzeria.Tests/OrderDataServiceTests.cs:
```diff
@@ -0,0 +1,203 @@
+            var resultList = await sut.GetMyOrdersAsync(user.Id);
+
+            // Assert
+            resultList.Should().NotBeNull();
```
not need for collections

Aleksander-Drumev-WA replied:
Removed

var1ap-WA commented on Pizzeria.Tests/OrderDataServiceTests.cs:
```diff
@@ -0,0 +1,203 @@
+            // Act
+            var orderId = await sut.CreateOrderAsync(basket.Id);
+
+            // Assert
```
need to check quantity on storage

Aleksander-Drumev-WA replied:
Done.

var1ap-WA commented on Pizzeria.Tests/OrderDataServiceTests.cs:
```diff
@@ -0,0 +1,203 @@
+            var orderId = await sut.CreateOrderAsync(basket.Id);
+
+            // Act
+            await sut.UpdateOrderStatusAsync(orderId, OrderStatus.Completed);
```
Is it possible?

Aleksander-Drumev-WA replied:
Replaced `enum` with string

var1ap-WA commented on Pizzeria.Tests/OrderDataServiceTests.cs:
```diff
@@ -0,0 +1,203 @@
+            // Assert
+            order.Should().NotBeNull();
+            order.OrderItems.All(oi => oi != null).Should().BeTrue();
+            order.OrderItems.Should().HaveCount(basketItems.Count);
```
https://fluentassertions.com/objectgraphs/

Aleksander-Drumev-WA replied:
I think I did it the right way now

var1ap-WA commented on WA.Pizza.Infrastructure/Migrations/20220218105717_NameAndPriceAddedToBasketItem.cs:
```diff
@@ -0,0 +1,37 @@
+            migrationBuilder.AddColumn<string>(
+                name: "Name",
+                table: "BasketItems",
+                type: "nvarchar(max)",
```
nvarchar(max)
be careful

Aleksander-Drumev-WA replied:
A `MaxLength` is set now

var1ap-WA commented on WA.Pizza.Infrastructure/Migrations/AppDbContextModelSnapshot.cs:
```diff
@@ -56,6 +56,13 @@ protected override void BuildModel(ModelBuilder modelBuilder)
 
+                    b.Property<string>("Name")
+                        .IsRequired()
+                        .HasColumnType("nvarchar(max)");
```
nvarchar(max)

Aleksander-Drumev-WA replied:
Changed

Aleksander-Drumev-WA committed:
Better graph comparison assertions.
Replacing for loops with foreach.
Minor fixes.

Aleksander-Drumev-WA committed:
Basket is now cleaned after Order creation.
